File: 2022/Dec2/Day2.py
# Part 1
Moves = [i.strip("\n").split() for i in open('input.txt', 'r').readlines()]
score  = {"X":1, "Y":2, "Z":3,
          "AX":3, "AY":6, "AZ":0,
          "BX":0, "BY":3, "BZ":6,
          "CX":6, "CY":0, "CZ":3}
print(sum([score[ i[1] ] + score[ i[0]+i[1] ] for i in Moves]))

# Part 2
move_needed={"AX":"C", "AY":"A", "AZ":"B",
             "BX":"A", "BY":"B", "BZ":"C",
             "CX":"B", "CY":"C", "CZ":"A"}
Moves2 = [i+[ move_needed[ i[0]+i[1]] ] for i in Moves]
score2 = {"A":1, "B":2, "C":3, "X":0, "Y":3, "Z":6}
print(sum([score2[i[1]]+score2[i[2]] for i in Moves2]))

# Version 2 using ord
Movs= [  i.strip('\n').split()  for i in open('input.txt', 'r').readlines()]
Moves=[ (ord(i[0])-64, ord(i[1])-87) for i in Movs  ]

def score(A,B):
    return (A-B+1)*3

tot=0
for B,A in Moves:
    if abs(A-B)<2:
        tot += A + score(A,B)
    elif(A<B):
        tot += A + score(A+3,B)
    else:
        tot += A + score(A,B+3)
print(tot)


# Smaller
Moves = [ (ord(i.strip('\n').split()[0])-64, ord(i.strip('\n').split()[1])-87) for i in open('input.txt', 'r').readlines()]
score=0
for B,A in Moves:
    score += A + (A-B+1)*3 if abs(A-B) <2 else ( A + (A+4-B)*3 if A<B else A+ (A-2-B)*3 )
print(score)

# One line
sum( [ A + (A-B+1)*3 if abs(A-B) <2 else ( A + (A+4-B)*3 if A<B else A+ (A-2-B)*3 ) for B,A in [(ord(i.strip('\n').split()[0])-64,ord(i.strip('\n').split()[1])-87) for i in open('input.txt', 'r').readlines()] ]  )

File: 2022/Dec2/Day2.scala
import scala.io.Source

val input :String = "input.txt"
val Lines = Source.fromFile(input).getLines().toList
val Moves = for(line<-Lines) yield List(line(0),line(2))

val score_me  = Map('X'->1,  'Y'->2,  'Z'->3,
                   "AX"->3, "AY"->6, "AZ"->0,
                   "BX"->0, "BY"->3, "BZ"->6,
                   "CX"->6, "CY"->0, "CZ"->3)


// Part 1
var my_score: Int =0
for(move<-Moves)
    my_score += score_me(move(1)) + score_me(s"${move(0)}${move(1)}")
println(my_score)

// Part 2
val move_needed=Map("AX"->'C', "AY"->'A', "AZ"->'B',
                    "BX"->'A', "BY"->'B', "BZ"->'C',
                    "CX"->'B', "CY"->'C', "CZ"->'A')
val Moves2 = for(move<-Moves) yield List(move(0),move(1), move_needed( s"${move(0)}${move(1)}") )

val score2 = Map('A'->1, 'B'->2, 'C'->3, 'X'->0, 'Y'->3, 'Z'->6)
my_score=0
for(move<-Moves2)
    my_score += score2(move(1))+score2(move(2))
println(my_score)
